마커 내보내기 (URL/JSON) 기능 구현 및 가져오기 (URL) 함수 추가

gitgitWi opened the pull request:
아래 TODO 내용과 같이, 
URL parsing 이후 `registerMarker`에서 `map` 값을 인식하지 못하는 문제가 있습니다
gif로 문제 상황 캡쳐했는데 파일 크기가 너무 커서 첨부가 안되서 일단 사진으로 첨부합니다

여기서만 몇시간 째 헤매고 있어서 일단 지금까지 한 내용만 PR로 보내고
JSON 가져오기 및 리팩토링은 되는대로 추가하도록 하겠습니다

---

# 구현 내용

- 수정된 타입 반영
- 일부 null/undefined 처리 연산에서 Nullish coalescing operator 사용

## JSON형식 내보내기에 마커 array 추가 

- 476ac21
- Export 관련 타입에 markers 추가; `ExportStyleMarkersType`
- 마커 내보내기 데이터는 instance 데이터에서 id/instance 제거하고 text/lng/lat만 내보냄

## URL형식 내보내기에 마커 Array 추가 

- 76385b4
- `urlParsing`에 `markerQueryString` 추가
- `isNotEmptyObject` 함수: `filteredStyle`, `markers` 비어있는 경우 url에 추가X

## URL에서 마커 정보 가져오기 함수 구현 

- 13551a0, b9b2932
- URL query string => parameter parsing 함수 구현
- `initializeMap` 함수 내 마커 등록 함수 추가

---

# TODO

## _URL에서 마커 정보 가져오는 부분_ => 일단 해결

- 13551a0
- b9b2932 에서 일단 해결되었습니다
  - `initializeMap` 함수에서 마커 등록하면서 `useMarkerFeature` hooks 전체가 참조하는 map과 다른 `map`을 참조했던 문제로 추정
  - `useMarkerFeature > useEffect` 안으로 해당 부분 옮겨서 해결

![image](https://user-images.githubusercontent.com/57997672/102018353-1e7d5200-3db0-11eb-9ee9-280740cc56f7.png)

![image](https://user-images.githubusercontent.com/57997672/102018426-859b0680-3db0-11eb-8bd0-7ef92e0e00fa.png)

- `registerMarker` 함수에 URL에서 parsing된 마커 정보 정상적으로 입력됨
- `useMarkerFeature` 내 `useSelector`로 가져온 `map`이 정상적으로 출력되지만
- `registerMarker` 내에서는 `map`의 값이`null`이 되는 문제로 마커가 등록되지 않음

Commit:
[FEAT] [#162] JSON형식 내보내기에 마커 array 추가

## 구현 내용

- Export 관련 타입에 markers 추가; `ExportStyleMarkersType`
- 마커 내보내기 데이터는 instance 데이터에서 id/instance 제거하고 text/lng/lat만 내보냄
- 일부 null/undefined 처리 연산에서 Nullish coalescing operator 사용

Commit:
[FEAT] [#162] URL형식 내보내기에 마커 Array 추가

- `urlParsing`에 `markerQueryString` 추가
- `isNotEmptyObject` 함수 선언: `filteredStyle`, `markers` 비어있는 경우 url에 추가X
- 수정된 타입 반영

Commit:
[FEAT] [#69] URL에서 마커 정보 가져오기 함수 구현

## 구현 내용

- URL query string => parameter parsing 함수 구현
- `initializeMap` 함수 내 마커 등록 함수 추가

## TODO

- 초기 map 초기화 이후 `useMarkerFeature` 내 `useSelector`로 가져온 `map`이 정상적으로 출력되지만
- `registerMarker` 내에서는 `map`의 값이`null`이 되는 문제로 마커가 등록되지 않음

Eunsol0410 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -165,30 +170,47 @@ function filterStyle(style: StoreDataType): StoreDataType {
+  >((state) => state) as any;
 
   const exportStyle = (): ExportType => {
     if (map || sidebar || history) {
```
marker도 조건문에 추가되면 좋을 것 같습니다~

gitgitWi replied:
네 이부분 추가로 push한 내용에서

```javascript
if (map || history || marker) {
````

이렇게 바뀌었습니다!

Eunsol0410 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -165,30 +170,47 @@ function filterStyle(style: StoreDataType): StoreDataType {
     return {
       filteredStyle: {},
       mapCoordinate: {},
+      markers: [],
```
빈배열로 마커를 내보내시는 이유가 있을까요?

gitgitWi replied:
여기도 리팩토링 전이라 `filteredStyle: {}, mapCoordinate: {},` 여기와 맞춰보느라 일단 빈배열로 보냈는데,
리팩토링하면서 `useExportStyleType`항목을 전부 옵셔널하게 바꾸어서 `return {}`으로 변경했습니다.

지금 보니까 아예 리턴타입에 null값을 추가해서 이 부분 아예 없애도 될 것 같아요

Eunsol0410 commented on src/components/Sidebar/SidebarModal/ExportModal.tsx:
```diff
@@ -53,26 +53,34 @@ const SubTitle = styled.h2`
+}: ExportType): string => {
   const data =
-    markers === [] ? { ...filteredStyle, markers } : { ...filteredStyle };
+    markers === [] ? { ...filteredStyle } : { ...filteredStyle, markers };
```
markers === [] 는 참조를 비교해서 항상 false가 되지 않을까요?
빈배열인지 체크하려고 하신다면 길이를 체크해도 좋을 것 같아요 😀

gitgitWi replied:
marker가 `undefined` 될 수 있다고 ts 에러가 떠서 바꾸다가 이렇게 했는데,
말씀하신대로 false만 나와서 리팩토링하면서 수정했습니다

지금 이부분은 `useExportModal`로 이동해놨어요

Eunsol0410 commented on src/store/marker/reducer.ts:
```diff
@@ -42,27 +42,34 @@ function markerReducer(
 
     case UPDATE_MARKER: {
-      const { id, lng, lat } = action.payload as MarkerType;
+      const { id, lng, lat } = action.payload as MarkerInstanceType;
```
Marker 스토어의 Instance는 맵박스 객체를 가리켜서 타입명이 조금 헷갈릴 수 있을 것 같은데, 어떻게 생각하시나요?
모든 Store상태 타입명을 일관되게 해주려고 MarkerType을 사용했었거든요 😁

gitgitWi replied:
`MarkerType`는 text, lng, lat를 갖고 있고,
`MarkerInstanceType`은 `MarkerType` 상속 +  id, instance를 갖고 있습니다.

`MarkerType`만 쓰이는 부분과 `MarkerInstanceType`이 쓰이는 부분이 있어서
편의상 나눴는데 네이밍은 좀더 가독성 좋은 것으로 해도 좋습니다!

Eunsol0410 replied:
아하! 이해되었어요~ 리팩토링 하면서 맞춰가면 좋을 것 같아요

Eunsol0410 commented on src/store/marker/action.ts:
```diff
@@ -42,24 +39,26 @@ export interface MarkerActionType {
+): MarkerActionType => ({
   type: ADD_MARKER,
-  payload: inputPayload,
+  payload: { ...inputPayload },
```
이렇게 수정하신 이유가 있을까요???

gitgitWi replied:
`inputPayload`가 객체로 전달되고 있어서, 
조금이라도 더 안전하게 깊은 복사가 될수 있도록 변경했습니다

Eunsol0410 commented on src/hooks/map/useMap.ts:
```diff
@@ -56,6 +56,11 @@ function useMap(): MapHookType {
+
+      states.markers?.forEach(({ lng, lat, text }) =>
+        registerMarker({ lngLat: { lng, lat }, text })
+      );
```
import에서 marker 리듀서를 업데이트 해야하지 않을까요??  어떤 용도로 사용하신건지 궁금합니다!
아직 marker 스토어가 초기화 되기 전이라서요..!

gitgitWi replied:
`/show` route로 들어왔을 때 설정해주는 부분이 이 부분이라서 여기에 넣었었는데,
다른 부분은 정상 작동하는데 registerMarker 만 잘 안되더라구요

크롬 디버거로 하나하나 다 찍어보니
이미 여러번 map 객체 로딩하고 함수를 실행하는데도 여기서만 map을 못찾고 null값으로 받고 있어서,
map에 대한 참조를 아예 다르게 하고 있는건가 추정하고 있습니다.

일단은 추가 push한 내용대로, `useMarkerFeature > useEffect` 안으로 넣어서 해결된 상태입니다

Commit:
[REFACTOR] [#69] `urlToJson...` 관련 함수 parsing 리팩토링

- queryString 전체 -> `parseQueryString` 값 사용
- `try-catch` 대신 empty string일 경우 early return
- 함수 return 타입 명시

Commit:
[REFACTOR] ExportModal hooks 이동

- component 코드인 `ExportModal`에 있는 로직을 `useExportStyle`로 이동
- `getStringifyStyleObject`에서 markers length/nullish 판단을 nullish 병합 연산으로 처리

Commit:
[FIX] [#69] URL에서 마커 가져오기 에러 해결 및 스타일 리턴타입 수정

## 해결 내용

- (13551a0)에서 `registerMarker`가 `map`을 정상적으로 참조하지 못하는 문제 해결
  - `initializeMap` 함수에서 마커 등록하면서 `useMarkerFeature` hooks 전체가 참조하는 map과 다른 `map`을 참조했던 문제로 추정
  - `useMarkerFeature > useEffect` 안으로 해당 부분 옮겨서 해결

- `urlParsing > urlToJsonGetStyleJson`에서 `styleParams` 없는 경우 `null` => 빈 객체 리턴으로 수정

- 일부 주석 및 `console.log` 제거

Commit:
[REFACTOR] marker reducer에서 localStorage 로직 분리 등

- marker reducer에서 로컬스토리지 로직 분리, 좀더 순수함수답게 조정
- router pathname 타입지정; `URLPathNameType`

qkrdmstlr3 commented on src/store/marker/reducer.ts:
```diff
@@ -8,84 +8,46 @@ import {
-      return { markers: newState };
-    }
+    case INIT_MARKER:
+      return { markers: [...(action.payload as MarkerInstanceType[])] };
```
상태관리와 연관없는 localStorage설정 부분을 뺐다고 보면 되겠죠??

gitgitWi replied:
네 reducer 내부를 최대한 순수하게 만들려고 `useMarkerFeature`로 뺐습니다

qkrdmstlr3 reviewed (commented):
대충 파악은 다 되었습니다! 감사합니다!

Commit:
Merge branch 'feature/history' of https://github.com/boostcamp-2020/Project08-A-Styled-Map-Admin-Tool into feature/marker

Co-authored-by: junhee lee <dlwnsgml94@gmail.com>
